ledger/smt_tools.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ledger
+
+// The Package Trie implements a sparse merkle trie.
+
+import (
+	"bytes"
+)
+
+// Get fetches the value of a key by going down the current trie root.
+func (s *SMT) Get(key []byte) ([]byte, error) {
+	return s.GetPreviousValue(s.Root, key)
+}
+
+// GetPreviousValue returns the value as of the specified root hash.
+func (s *SMT) GetPreviousValue(prevRoot []byte, key []byte) ([]byte, error) {
+	s.lock.RLock()
```

Comment:
Is a read lock correct, given the code is writing to s.atomicUpdate below?

Reply:
I wondered the same thing.  I have a test suite that replaces the lock with a test mock that never blocks, and am running a few simulations now to determine if this is going to cause issues.  I will let you know how it goes.